File: deno/lib/oss/oss.router.ts
import {
  Router,
  type RouterContext,
} from "@oak/oak";

import {
  error,
} from "/lib/context.ts";

// import {
//   resize,
// } from "/lib/image/mod.ts";

import {
  handleRequestId,
} from "/lib/oak/request_id.ts";

const router = new Router({
  prefix: "/api/oss/",
});

router.post("upload", async function(ctx) {
  const request = ctx.request;
  const response = ctx.response;
  if (await handleRequestId(
    response,
    request.headers.get("x-request-id"),
  )) {
    return;
  }
  const body = request.body;
  const contentType = body.type().toLocaleLowerCase();
  if (!request.hasBody || contentType !== "form-data") {
    response.status = 415;
    return;
  }
  const formData = await body.formData();
  const file = formData.get("file");
  if (!file || !(file instanceof File)) {
    throw new Error("file must a form-data file!");
  }
  const {
    upload
  } = await import("./oss.service.ts");
  const id = await upload(file);
  response.body = {
    code: 0,
    data: id,
  };
});

// deno-lint-ignore no-explicit-any
async function download(ctx: RouterContext<any>) {
  const request = ctx.request;
  const response = ctx.response;
  const {
    statObject,
    getObject,
  } = await import("./oss.service.ts");
  try {
    let filename: string | undefined = ctx.params?.filename;
    const searchParams = request.url.searchParams;
    let inline = searchParams.get("inline");
    const id = searchParams.get("id")!;
    if (!inline) {
      inline = "1";
    }
    let attachment = "inline";
    if (inline != "1") {
      attachment = "attachment";
    }
    if (filename) {
      filename = encodeURIComponent(filename);
    }
    const stats = await statObject(id);
    if (stats) {
      if (stats.contentType) {
        response.headers.set("Content-Type", stats.contentType);
      }
      if (!filename && stats?.meta?.filename) {
        filename = stats?.meta?.filename;
      }
      if (stats.contentLength) {
        response.headers.set("Content-Length", stats.contentLength.toString());
      }
      if (stats.etag) {
        response.headers.set("ETag", stats.etag);
      }
      if (request.headers.get("if-none-match") === stats.etag) {
        response.headers.set("Content-Disposition", `${ attachment }; filename=${ filename || encodeURIComponent(id) }`);
        response.status = 304;
        return;
      }
    }
    response.headers.set("Content-Disposition", `${ attachment }; filename=${ filename || encodeURIComponent(id) }`);
    const objInfo = await getObject(id);
    if (!objInfo || !objInfo.body) {
      const err = new Error("NotFound");
      // deno-lint-ignore no-explicit-any
      (err as any).code = "NotFound";
      throw err;
    }
    const res = new Response(objInfo.body);
    response.body = await res.arrayBuffer();
  } catch (err0) {
    const err = err0 as Error;
    // deno-lint-ignore no-explicit-any
    if ((err as any).code === "NotFound") {
      const errMsg = "文件不存在!";
      response.status = 404;
      response.body = errMsg;
      return;
    }
    error(err);
    const errMsg = err?.message || err?.toString() || err || "";
    response.status = 500;
    response.body = errMsg;
    return;
  }
}

// deno-lint-ignore no-explicit-any
router.get("download/:filename", <any>download);
// deno-lint-ignore no-explicit-any
router.get("download", <any>download);

router.get("img", async function(ctx) {
  const request = ctx.request;
  const response = ctx.response;
  const {
    statObject,
    getObject,
  } = await import("./oss.service.ts");
  try {
    let filename: string | undefined = ctx.params?.filename;
    const searchParams = request.url.searchParams;
    let inline = searchParams.get("inline");
    const id = searchParams.get("id")!;
    const f = searchParams.get("f");
    const w = searchParams.get("w");
    const h = searchParams.get("h");
    const q = searchParams.get("q");
    if (!inline) {
      inline = "1";
    }
    let attachment = "inline";
    if (inline != "1") {
      attachment = "attachment";
    }
    if (filename) {
      filename = encodeURIComponent(filename);
    }
    const stats = await statObject(id);
    if (!stats) {
      const err = new Error("NotFound");
      // deno-lint-ignore no-explicit-any
      (err as any).code = "NotFound";
      throw err;
    }
    if (stats.contentType) {
      response.headers.set("Content-Type", stats.contentType);
    }
    if (!filename && stats?.meta?.filename) {
      filename = stats?.meta?.filename;
    }
    if (stats.contentLength > 0) {
      response.headers.set("Content-Length", stats.contentLength.toString());
    }
    if (stats.etag) {
      response.headers.set("ETag", stats.etag);
    }
    if (request.headers.get("if-none-match") === stats.etag) {
      response.headers.set("Content-Disposition", `${ attachment }; filename=${ filename || encodeURIComponent(id) }`);
      response.status = 304;
      return;
    }
    if (!f) {
      const objInfo = await getObject(id);
      if (!objInfo || !objInfo.body) {
        const err = new Error("NotFound");
        // deno-lint-ignore no-explicit-any
        (err as any).code = "NotFound";
        throw err;
      }
    }
    let is_img = false;
    if (stats.contentType) {
      const contentType = stats.contentType.toLowerCase();
      if (contentType.startsWith("image/") && !contentType.startsWith("image/svg")) {
        is_img = true;
      }
    }
    if (!is_img) {
      const objInfo = await getObject(id);
      if (!objInfo || !objInfo.body) {
        const err = new Error("NotFound");
        // deno-lint-ignore no-explicit-any
        (err as any).code = "NotFound";
        throw err;
      }
      response.body = objInfo.body;
      return;
    }
    // 如果缩略图已经存在, 则直接返回
    let cache_id = id;
    if (f) {
      cache_id += `-f${ f }`;
    }
    if (w) {
      cache_id += `-w${ w }`;
    }
    if (h) {
      cache_id += `-h${ h }`;
    }
    if (q) {
      cache_id += `-q${ q }`;
    }
    const {
      statObject: statObjectTmpfile,
      getObject: getObjectTmpfile,
      putObject: putObjectTmpfile,
    } = await import("/lib/tmpfile/tmpfile.service.ts");
    const img_stats = await statObjectTmpfile(cache_id);
    if (img_stats) {
      if (img_stats.contentType) {
        response.headers.set("Content-Type", img_stats.contentType);
      }
      if (!filename && img_stats?.meta?.filename) {
        filename = img_stats?.meta?.filename;
      }
      if (img_stats.contentLength) {
        response.headers.set("Content-Length", img_stats.contentLength.toString());
      }
      if (img_stats.etag) {
        response.headers.set("ETag", img_stats.etag);
      }
      if (request.headers.get("if-none-match") === img_stats.etag) {
        response.headers.set("Content-Disposition", `${ attachment }; filename=${ filename || encodeURIComponent(id) }`);
        response.status = 304;
        return;
      }
      response.headers.set("Content-Disposition", `${ attachment }; filename=${ filename || encodeURIComponent(id) }`);
      const objInfo = await getObjectTmpfile(cache_id);
      if (!objInfo || !objInfo.body) {
        const err = new Error("NotFound");
        // deno-lint-ignore no-explicit-any
        (err as any).code = "NotFound";
        throw err;
      }
      const res = new Response(objInfo.body);
      const content = await res.arrayBuffer();
      response.headers.set("Content-Length", content.byteLength.toString());
      response.body = content;
      return;
    }
    const objInfo = await getObject(id);
    if (!objInfo || !objInfo.body) {
      const err = new Error("NotFound");
      // deno-lint-ignore no-explicit-any
      (err as any).code = "NotFound";
      throw err;
    }
    const res = new Response(objInfo.body);
    const content = new Uint8Array(await res.arrayBuffer());
    const format = f || "webp"; 
    const height = h == null ? undefined : parseInt(h);
    const width = w == null ? undefined : parseInt(w);
    const quality = q == null ? undefined : parseInt(q);
    
    const {
      resize,
    } = await import("/lib/image/mod.ts");
    
    const content2 = resize(
      content,
      format,
      width,
      height,
      quality,
    );
    
    const len = content2.byteLength;
    response.headers.set("Content-Length", len.toString());
    
    let content_type = "image/webp";
    if (format === "jpeg") {
      content_type = "image/jpeg";
    } else if (format === "png") {
      content_type = "image/png";
    } else if (format === "webp") {
      content_type = "image/webp";
    }
    response.headers.set("Content-Type", content_type);
    
    // 修改filename后缀名
    if (filename) {
      const arr = filename.split(".");
      if (arr.length > 1) {
        arr.pop();
      }
      filename = arr.join(".");
      filename += `.${ format }`;
    }
    response.headers.set("Content-Disposition", `${ attachment }; filename=${ filename || encodeURIComponent(id) }`);
    
    response.headers.set("Content-Type", content_type);
    
    // 缓存图片到tmpfile
    await putObjectTmpfile(
      cache_id,
      content2,
      content_type,
    );
    
    const img_stats2 = await statObjectTmpfile(cache_id);
    if (img_stats2) {
      if (img_stats2.etag) {
        response.headers.set("ETag", img_stats2.etag);
      }
      if (request.headers.get("if-none-match") === img_stats2.etag) {
        response.status = 304;
        return;
      }
    }
    
    response.body = content2;
    return;
  } catch (err0) {
    const err = err0 as Error;
    // deno-lint-ignore no-explicit-any
    if ((err as any).code === "NotFound") {
      const errMsg = "文件不存在!";
      response.status = 404;
      response.body = errMsg;
      return;
    }
    error(err);
    const errMsg = err?.message || err?.toString() || err || "";
    response.status = 500;
    response.body = errMsg;
    return;
  }
});

export default router;

File: rust/src/common/oss/oss_router.rs
use anyhow::Result;

use poem::{
  handler, web::{Multipart, Path, Query}, Response,
};
use poem::http::StatusCode;
use serde::{Deserialize, Serialize};
use serde_json::json;

use std::io::Cursor;
use image::ImageReader;
use image::ImageFormat;
use image::imageops::FilterType;
use image::GenericImageView;

use crate::common::{context::get_short_uuid, tmpfile::tmpfile_dao};

use super::oss_service;

#[handler]
pub async fn upload(
  mut multipart: Multipart,
) -> Result<Response> {
  let mut file_name = String::new();
  let mut content_type: Option<String> = None;
  let mut content: Option<Vec<u8>> = None;
  while let Ok(Some(field)) = multipart.next_field().await {
    let name = field.name();
    if name != Some("file") {
      continue;
    }
    file_name = field.file_name().map(ToString::to_string).unwrap_or("".to_owned());
    content_type = field.content_type().map(ToString::to_string);
    if let Ok(bytes) = field.bytes().await {
      content = Some(bytes.to_vec());
      break;
    }
  }
  if content.is_none() {
    let mut response = Response::builder();
    response = response.header("Content-Type", "application/json");
    let response = response.body(json!({
      "code": 1,
      "msg": "上传失败",
      "data": null,
    }).to_string());
    return Ok(response);
  }
  let content = content.unwrap();
  let content_type = content_type.unwrap_or("application/octet-stream".to_owned());
  let id = get_short_uuid();
  oss_service::put_object(id.as_str(), &content, &content_type, &file_name).await?;
  let mut response = Response::builder();
  response = response.header("Content-Type", "application/json");
  let response = response.body(json!({
    "code": 0,
    "data": id,
  }).to_string());
  Ok(response)
}

#[handler]
pub async fn delete(
  Query(DownloadQuery { id, inline: _inline }): Query<DownloadQuery>,
) -> Result<String> {
  let res = oss_service::delete_object(&id).await?;
  let res = json!({
    "code": 0,
    "data": res,
  }).to_string();
  Ok(res)
}

#[derive(Deserialize)]
pub struct DownloadQuery {
  id: String,
  inline: Option<String>,
}

async fn _download(
  mut filename: String,
  id: String,
  inline: Option<String>,
  req: &poem::Request,
) -> Result<Response> {
  if id.is_empty() {
    return Ok(Response::builder().status(StatusCode::from_u16(404)?).finish());
  }
  let if_none_match = req.header("if-none-match");
  let inline = inline.unwrap_or("1".to_owned());
  let attachment = match inline.as_str() {
    "1" => "inline",
    _ => "attachment",
  };
  let stat = oss_service::head_object(&id).await?;
  if stat.is_none() {
    return Ok(Response::builder().status(StatusCode::from_u16(404)?).finish());
  }
  let stat = stat.unwrap();
  let mut response = Response::builder();
  if let Some(content_length) = stat.content_length {
    if content_length <= 0 {
      return Ok(Response::builder().status(StatusCode::from_u16(404)?).finish());
    }
  }
  if let Some(content_type) = &stat.content_type {
    if !content_type.is_empty() {
      response = response.content_type(content_type);
    }
  }
  filename = filename.trim().to_string();
  if filename.is_empty() {
    filename = stat.filename;
    if filename.is_empty() {
      filename = urlencoding::encode(id.as_str()).to_string();
    }
  } else {
    filename = urlencoding::encode(filename.as_str()).to_string();
  }
  response = response.header("Content-Disposition", format!("{attachment}; filename={filename}"));
  if let Some(last_modified) = &stat.last_modified {
    if !last_modified.is_empty() {
      response = response.header("Last-Modified", last_modified);
    }
  }
  if let Some(etag) = &stat.etag {
    if !etag.is_empty() {
      response = response.header("ETag", etag);
    }
  }
  if let Some(if_none_match) = if_none_match {
    if let Some(etag) = stat.etag {
      if if_none_match == etag {
        response = response.status(StatusCode::from_u16(304)?);
        return Ok(response.finish());
      }
    }
  }
  let content = oss_service::get_object(&id).await?;
  response = response.header("Content-Length", content.len().to_string());
  let response = response.body(content);
  Ok(response)
}

#[handler]
pub async fn download_filename(
  Path(mut filename): Path<String>,
  Query(DownloadQuery { id, inline }): Query<DownloadQuery>,
  req: &poem::Request,
) -> Result<Response> {
  _download(filename, id, inline, req).await
}

#[handler]
pub async fn download(
  Query(DownloadQuery { id, inline }): Query<DownloadQuery>,
  req: &poem::Request,
) -> Result<Response> {
  let filename = "".to_owned();
  _download(filename, id, inline, req).await
}

#[derive(Deserialize, Serialize)]
pub struct ImgQuery {
  id: String,
  inline: Option<String>,
  f: Option<String>,
  w: Option<u32>,
  h: Option<u32>,
  q: Option<u8>,
}

#[handler]
pub async fn img(
  Query(ImgQuery {
    id,
    inline,
    f,
    w,
    h,
    q,
  }): Query<ImgQuery>,
  req: &poem::Request,
) -> Result<Response> {
  if id.is_empty() {
    return Ok(Response::builder().status(StatusCode::from_u16(404)?).finish());
  }
  let if_none_match = req.header("if-none-match");
  let mut filename = "".to_owned();
  let inline = inline.unwrap_or("1".to_owned());
  let attachment = match inline.as_str() {
    "1" => "inline",
    _ => "attachment",
  };
  let stat = oss_service::head_object(&id).await?;
  if stat.is_none() {
    return Ok(Response::builder().status(StatusCode::from_u16(404)?).finish());
  }
  let stat = stat.unwrap();
  let mut response = Response::builder();
  if let Some(content_length) = stat.content_length {
    if content_length <= 0 {
      return Ok(Response::builder().status(StatusCode::from_u16(404)?).finish());
    }
  }
  if let Some(content_type) = &stat.content_type {
    if !content_type.is_empty() {
      response = response.content_type(content_type);
    }
  }
  filename = filename.trim().to_string();
  let file_name_empty = filename.is_empty();
  if file_name_empty {
    filename = stat.filename;
    if filename.is_empty() {
      filename = urlencoding::encode(id.as_str()).to_string();
    }
  } else {
    filename = urlencoding::encode(filename.as_str()).to_string();
  }
  let cache_id = {
    let mut id = id.clone();
    if let Some(f) = &f {
      id.push_str("-f");
      id.push_str(f);
    }
    if let Some(w) = w {
      id.push_str("-w");
      id.push_str(w.to_string().as_str());
    }
    if let Some(h) = h {
      id.push_str("-h");
      id.push_str(h.to_string().as_str());
    }
    if let Some(q) = q {
      id.push_str("-q");
      id.push_str(q.to_string().as_str());
    }
    id
  };
  let f_is_none = f.is_none();
  let format = f.unwrap_or("webp".to_owned());
  let output_format = match format.as_str() {
    "webp" => ImageFormat::WebP,
    "jpg" => ImageFormat::Jpeg,
    "png" => ImageFormat::Png,
    "ico" => ImageFormat::Ico,
    _ => ImageFormat::WebP,
  };
  let content_type = match format.as_str() {
    "webp" => "image/webp",
    "jpg" => "image/jpeg",
    "png" => "image/png",
    "ico" => "image/x-icon",
    _ => "image/webp",
  };
  // 修改filename后缀名
  let mut filename_vec: Vec<&str> = filename.split('.').collect();
  if filename_vec.len() > 1 {
    filename_vec.pop();
    filename = filename_vec.join(".");
    filename.push('.');
    filename.push_str(format.as_str());
    if filename.contains('"') {
      filename = filename.replace('"', "");
    }
  }
  if let Some(if_none_match) = if_none_match {
    if let Some(etag) = &stat.etag {
      if if_none_match == etag {
        if let Some(last_modified) = &stat.last_modified {
          if !last_modified.is_empty() {
            response = response.header("Last-Modified", last_modified);
          }
        }
        if !etag.is_empty() {
          response = response.header("ETag", etag);
        }
        response = response.header("Content-Disposition", format!("{attachment}; filename=\"{filename}\""));
        response = response.status(StatusCode::from_u16(304)?);
        return Ok(response.finish());
      } else if !etag.is_empty() {
        response = response.header("ETag", etag);
      }
    }
  }
  let is_img = match &stat.content_type {
    Some(content_type) => {
      content_type.starts_with("image/") && !content_type.starts_with("image/svg")
    },
    None => false,
  };
  if f_is_none || !is_img {
    response = response.header("Content-Disposition", format!("{attachment}; filename=\"{filename}\""));
    let content = oss_service::get_object(&id).await?;
    let len = content.len();
    response = response.header("Content-Length", len.to_string());
    if let Some(last_modified) = &stat.last_modified {
      if !last_modified.is_empty() {
        response = response.header("Last-Modified", last_modified);
      }
    }
    if let Some(etag) = &stat.etag {
      if !etag.is_empty() {
        response = response.header("ETag", etag);
      }
    }
    let response = response.body(content);
    return Ok(response);
  }
  
  let stat = tmpfile_dao::head_object(&cache_id).await?;
  if let Some(stat) = stat {
    if let Some(content_type) = &stat.content_type {
      if !content_type.is_empty() {
        response = response.content_type(content_type);
      }
    }
    if file_name_empty {
      filename = stat.filename;
      if filename.is_empty() {
        filename = urlencoding::encode(cache_id.as_str()).to_string();
      }
    } else {
      filename = urlencoding::encode(filename.as_str()).to_string();
    }
    if let Some(if_none_match) = if_none_match {
      if let Some(etag) = &stat.etag {
        if if_none_match == etag {
          if let Some(last_modified) = &stat.last_modified {
            if !last_modified.is_empty() {
              response = response.header("Last-Modified", last_modified);
            }
          }
          if !etag.is_empty() {
            response = response.header("ETag", etag);
          }
          response = response.header("Content-Disposition", format!("{attachment}; filename=\"{filename}\""));
          response = response.status(StatusCode::from_u16(304)?);
          return Ok(response.finish());
        } else if !etag.is_empty() {
          response = response.header("ETag", etag);
        }
      }
    }
    if let Some(etag) = &stat.etag {
      if !etag.is_empty() {
        response = response.header("ETag", etag);
      }
    }
    let content = tmpfile_dao::get_object(&cache_id).await?;
    response = response.header("Content-Length", content.len().to_string());
    response = response.header("Content-Disposition", format!("{attachment}; filename=\"{filename}\""));
    response = response.content_type(content_type);
    let response = response.body(content);
    return Ok(response);
  }
  let content = oss_service::get_object(&id).await?;
  let len = content.len();
  let img = ImageReader::new(Cursor::new(&content)).with_guessed_format()?.decode()?;
  let (width, height) = img.dimensions();
  let mut nwidth = w.unwrap_or(width);
  let mut nheight = h.unwrap_or(height);
  if nwidth > width {
    nwidth = width;
  }
  if nheight > height {
    nheight = height;
  }
  // if nwidth > width || nheight > height {
  //   response = response.header("Content-Length", len.to_string());
  //   let response = response.body(content);
  //   return Ok(response);
  // }
  drop(content);
  let img = img.resize(nwidth, nheight, FilterType::Lanczos3);
  // let img = img.into_rgba8();
  let mut content: Vec<u8> = Vec::with_capacity(len);
  img.write_to(&mut Cursor::new(&mut content), output_format)?;
  
  let len = content.len();
  
  // 缓存图片到tmpfile
  tmpfile_dao::put_object(
    &cache_id,
    &content,
    content_type,
    &filename,
  ).await?;
  
  let stat = tmpfile_dao::head_object(&cache_id).await?;
  
  if let Some(stat) = &stat {
    if let Some(if_none_match) = if_none_match {
      if let Some(etag) = &stat.etag {
        if if_none_match == etag {
          if let Some(last_modified) = &stat.last_modified {
            if !last_modified.is_empty() {
              response = response.header("Last-Modified", last_modified);
            }
          }
          if !etag.is_empty() {
            response = response.header("ETag", etag);
          }
          response = response.header("Content-Length", len.to_string());
          response = response.content_type(content_type);
          response = response.header("Content-Disposition", format!("{attachment}; filename=\"{filename}\""));
          response = response.status(StatusCode::from_u16(304)?);
          return Ok(response.finish());
        } else if !etag.is_empty() {
          response = response.header("ETag", etag);
        }
      }
    }
  }
  if let Some(stat) = stat {
    if let Some(etag) = &stat.etag {
      if !etag.is_empty() {
        response = response.header("ETag", etag);
      }
    }
  }
  response = response.header("Content-Length", len.to_string());
  response = response.content_type(content_type);
  response = response.header("Content-Disposition", format!("{attachment}; filename=\"{filename}\""));
  let response = response.body(content);
  Ok(response)
}
